xen/arch/arm/vscmi.c

```diff
@@ -0,0 +1,356 @@
+        write_32(data->msg_payload + 4, SCMI_PROTO_ATTRS);
+        data->length = sizeof(uint32_t) * 3;
+        break;
+    case BASE_DISCOVER_VENDOR:
+        write_32(data->msg_payload, SCMI_SUCCESS);
+        strlcpy((char*)data->msg_payload + 4, SCMI_VENDOR, SCMI_MAX_STR_SIZE);
+        data->length = sizeof(uint32_t) * 2 + sizeof(SCMI_VENDOR);
+        break;
+    case BASE_DISCOVER_SUB_VENDOR:
+        write_32(data->msg_payload, SCMI_SUCCESS);
+        strlcpy((char*)data->msg_payload + 4, SCMI_SUBVENDOR, SCMI_MAX_STR_SIZE);
+        data->length = sizeof(uint32_t) * 2 + sizeof(SCMI_SUBVENDOR);
+        break;
+    case BASE_DISCOVER_LIST_PROTOCOLS:
+    {
+        uint32_t skip = *(uint32_t*)data->msg_payload;
+        if ( skip == 0 ) {
+            write_32(data->msg_payload, SCMI_SUCCESS);
+            write_32(data->msg_payload + 4, 1);
+            write_32(data->msg_payload + 8, SCMI_PROTOCOL_PERF);
+            data->length = sizeof(uint32_t) * 4;
+        } else {
+            write_32(data->msg_payload, SCMI_ERR_RANGE);
+            data->length = sizeof(uint32_t) * 2;
+        }
+
+        break;
+    }
+    case BASE_DISCOVER_AGENT:
+        write_32(data->msg_payload, SCMI_SUCCESS);
+        strlcpy((char*)data->msg_payload + 4, SCMI_AGENT, SCMI_MAX_STR_SIZE);
+        data->length = sizeof(uint32_t) * 2 + sizeof(SCMI_AGENT);
+        break;
+    default:
+        write_32(data->msg_payload, SCMI_ERR_SUPPORT);
+        data->length = sizeof(uint32_t) * 2;
+        break;
+    }
+}
+
+static void handle_perf_req(struct scmi_shared_mem *data)
+{
+    switch ( SCMI_HDR_MSG_ID(data->msg_header) )
+    {
+    case PROTOCOL_VERSION:
+        write_32(data->msg_payload, SCMI_SUCCESS);
+        write_32(data->msg_payload + 4, SCMI_VERSION);
+        data->length = sizeof(uint32_t) * 3;
+        break;
+    case PROTOCOL_ATTRIBUTES:
+    {
+        struct scmi_msg_resp_perf_attributes* attrs =
+            (struct scmi_msg_resp_perf_attributes*)(data->msg_payload + 4);
+        write_32(data->msg_payload, SCMI_SUCCESS);
+
+        attrs->num_domains = cpu_to_le16(current->domain->max_vcpus);
+        attrs->flags = cpu_to_le16(0);
+        attrs->stats_addr_low = cpu_to_le32(0);
+        attrs->stats_addr_high = cpu_to_le32(0);
+        attrs->stats_size = cpu_to_le32(0);
+        data->length = sizeof(*attrs) + sizeof(uint32_t) * 2;
+        break;
+    }
+    case PERF_DOMAIN_ATTRIBUTES:
+    {
+        uint32_t domain_id = le32_to_cpu(*(__le32*)data->msg_payload);
+        struct scmi_msg_resp_perf_domain_attributes* attrs =
+            (struct scmi_msg_resp_perf_domain_attributes*)(data->msg_payload + 4);
+
+        if ( domain_id >= current->domain->max_vcpus )
+        {
+            write_32(data->msg_payload, SCMI_ERR_RANGE);
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+        write_32(data->msg_payload, SCMI_SUCCESS);
+
+        attrs->flags = cpu_to_le32(BIT(31) | BIT(30));
```

Comment:
Common to handle_base_req and handle_perf_req:
Please  minimize the usage of different magics as possible.